### PR TITLE
import settings changes for importing code

### PR DESCRIPTION
**Pull request by DhaaraniCIT**

### CODE REVIEW

File: src/app/core/models/common/import-settings.model.ts [hunk 1]
```diff
@@ -52,11 +54,19 @@ export class ImportSettingsModel {
       destination_field: new FormControl(data.destination_field || '', RxwebValidators.unique()),
       import_to_fyle: new FormControl(data.import_to_fyle || false),
       is_custom: new FormControl(data.is_custom || false),
-      source_placeholder: new FormControl(data.source_placeholder || null)
+      source_placeholder: new FormControl(data.source_placeholder || null),
+      import_code: new FormControl(data.import_code || false, RxwebValidators.unique())
```
Comment (ashwin1111) [inline src/app/core/models/common/import-settings.model.ts:58]: why do we need unique validator here? what case do we get dups here

File: src/app/core/models/common/import-settings.model.ts [hunk 2]
```diff
@@ -52,11 +54,19 @@ export class ImportSettingsModel {
...
     });
   }
 
-  static constructFormArray(importSettingsMappingSettings: ImportSettingMappingRow[] | [], accountingAppFields: IntegrationField[], isDestinationFixedImport: boolean = true): FormGroup[] {
+  static getImportCodeField(importCodeFields: string[] | [], destinationField: string): boolean {
```
Comment (ashwin1111) [inline src/app/core/models/common/import-settings.model.ts:62]: added comment to previous PR

File: src/app/core/models/common/import-settings.model.ts [hunk 3]
```diff
@@ -72,9 +82,11 @@ export class ImportSettingsModel {
           import_to_fyle: false,
           is_custom: false,
           source_field: '',
-          source_placeholder: null
+          source_placeholder: null,
+          import_code: this.getImportCodeField(importCodeFields, accountingAppField.attribute_type)
       };
       if (mappingSetting) {
+        fieldData.import_code = this.getImportCodeField(importCodeFields, accountingAppField.attribute_type);
```
Comment (ashwin1111) [inline src/app/core/models/common/import-settings.model.ts:89]: why is this required?

File: src/app/core/models/sage300/sage300-configuration/sage300-import-settings.model.ts [hunk 4]
```diff
@@ -51,10 +53,14 @@ export class Sage300ImportSettingModel extends ImportSettingsModel {
     }
 
     static mapAPIResponseToFormGroup(importSettings: Sage300ImportSettingGet | null, sage300Fields: IntegrationField[]): FormGroup {
-        const expenseFieldsArray = importSettings?.mapping_settings ? this.constructFormArray(importSettings.mapping_settings, sage300Fields, false) : [] ;
+        const import_code = importSettings?.import_settings?.import_code_fields ? importSettings?.import_settings?.import_code_fields : [];
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-import-settings.model.ts:56]: snake case

File: src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts [hunk 5]
```diff
@@ -66,12 +66,14 @@ export class Sage300ImportSettingsComponent implements OnInit {
     {
       source_field: DefaultImportFields.CATEGORY,
       destination_field: DefaultImportFields.ACCOUNT,
-      formController: 'importCategories'
+      formController: 'importCategories',
+      import_code: 'importCategoryCode'
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts:70]: didn't we plan to make it true/false during ED?

File: src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts [hunk 6]
```diff
@@ -284,6 +286,20 @@ export class Sage300ImportSettingsComponent implements OnInit {
...
+	        fields.push(value.destination_field);
+	        this.importSettingForm.get('importCodeFields')?.patchValue(fields);
+        } else if (!control.value.import_code || !control.value.import_to_fyle) {
+          const onboardingState = this.workspaceService.getOnboardingState();
+	        if (onboardingState === Sage300OnboardingState.IMPORT_SETTINGS) {
+            const fields = this.importSettingForm.get('importCodeFields')?.value;
+            if (fields.length > 0) {
+              const index = fields.indexOf(value.destination_field);
+              index > -1 ? fields.splice(index, 1) : '';
+            }
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts:300]: simplify code

File: src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts [hunk 7]
```diff
@@ -337,9 +353,69 @@ export class Sage300ImportSettingsComponent implements OnInit {
...
+  private defaultFieldWatcher() {
+    this.importSettingForm.controls.importCategories.valueChanges.subscribe((importFromFyle) => {
+      if (!importFromFyle) {
+        this.importSettingForm.controls.importCategoryCode.patchValue(false);
+        const fields = this.importSettingForm.get('importCodeFields')?.value;
+	      if (fields.length > 0) {
+		      const index = fields.indexOf(DefaultImportFields.ACCOUNT);
+		      index > -1 ? fields.splice(index, 1) : '';
+	      }
+        this.importSettingForm.get('importCodeFields')?.patchValue(fields);
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts:365]: all these are duplicated logic, move to common place

File: src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts [hunk 8]
```diff
@@ -337,9 +353,69 @@ export class Sage300ImportSettingsComponent implements OnInit {
...
+      }
+    });
+
+    this.importSettingForm.controls.importVendorAsMerchant.valueChanges.subscribe((importFromFyle) => {
+      if (!importFromFyle) {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts:370]: same

File: src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts [hunk 9]
```diff
@@ -337,9 +353,69 @@ export class Sage300ImportSettingsComponent implements OnInit {
...
+	      }
+        this.importSettingForm.get('importCodeFields')?.patchValue(fields);
+      }
+    });
+  }
+
+  private defaultFieldCodeImportFieldWatcher() {
+    this.importSettingForm.controls.importCategoryCode.valueChanges.subscribe((importCode) => {
+      if (importCode && this.importSettingForm.controls.importCategories) {
+        const fields = this.importSettingForm.get('importCodeFields')?.value;
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts:385]: same, add comments to code

File: src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts [hunk 10]
```diff
@@ -337,9 +353,69 @@ export class Sage300ImportSettingsComponent implements OnInit {
...
+		      const index = fields.indexOf(DefaultImportFields.ACCOUNT);
+		      index > -1 ? fields.splice(index, 1) : '';
+	      }
+        this.importSettingForm.get('importCodeFields')?.patchValue(fields);
+      }
+    });
+
+    this.importSettingForm.controls.importVendorCode.valueChanges.subscribe((importCode) => {
+      if (importCode && this.importSettingForm.controls.importVendorAsMerchant) {
+        const fields = this.importSettingForm.get('importCodeFields')?.value;
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts:400]: same as above

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html [hunk 11]
```diff
@@ -99,8 +99,9 @@ <h5 *ngIf="!isCloneSettingView" class="!tw-text-text-muted tw-text-14-px !tw-fon
                         {{brandingConfig.brandName}} Fields should be unique
                         </p>
                     </div>
-                <p-inputSwitch *ngIf="!isDestinationFixedImport" (onChange)="onSwitchChanged($event, expenseField)" [pTooltip]="isExpenseFieldDependent(expenseField.value) ? 'You cannot import dependent fields' : '' " (click)="updateDependentField(expenseField.value.source_field,expenseField.value.import_to_fyle)" formControlName="import_to_fyle" class="input-toggle-section tw-pl-4-px tw-pt-10-px"></p-inputSwitch>
+                <p-inputSwitch *ngIf="!isDestinationFixedImport" (onChange)="onSwitchChanged($event, expenseField)" [pTooltip]="isExpenseFieldDependent(expenseField.value) ? 'You cannot import dependent fields' : (!expenseField.value.source_field) ? helper.sentenseCaseConversion('Fill in the Fyle Field toggle will be enabled') : '' " (click)="updateDependentField(expenseField.value.source_field,expenseField.value.import_to_fyle)" formControlName="import_to_fyle" class="input-toggle-section tw-pt-10-px" [disabled]="!expenseField.value.source_field"></p-inputSwitch>
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html:102]: move tooltip generation to ts, too much conditions

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html [hunk 12]
```diff
@@ -22,7 +22,7 @@
                     <input type="text" class="tw-text-14-px tw-text-text-secondary !tw-font-500 tw-w-300-px !tw-px-14-px !tw-py-10-px tw-border tw-border-solid !tw-border-border-tertiary tw-rounded-4-px" value="{{ defaultField.source_field | snakeCaseToSpaceCase | titlecase }}" disabled>
                 </div>
                 <div class="input-toggle-section tw-w-10-vw tw-pt-10-px">
-                    <p-inputSwitch [formControlName]="defaultField.formController"></p-inputSwitch>
+                    <p-inputSwitch [formControlName]="defaultField.formController" (onChange)="onShowWarningForDependentFields($event, form)"></p-inputSwitch>
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html:25]: rename func since it's going to more job now, and not just related to dependent fields

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html [hunk 13]
```diff
@@ -170,5 +171,58 @@ <h5 *ngIf="!isCloneSettingView" class="!tw-text-text-muted tw-text-14-px !tw-fon
...
+                    ></app-configuration-select-field>
+                </div>
+            </div>
+            <div formArrayName="expenseFields">
+                <div *ngFor="let expenseField of expenseFieldsGetter.controls; let i = index" [formGroupName]="i">
+                    <div *ngIf="expenseField.value.import_to_fyle">
+                        <app-configuration-select-field
+                        [appName]="appName"
+                        [form]="getFormGroup(expenseField)"
+                        [options]="importCodeSelectorOptions"
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html:215]: check with Kamalini, if we should do different example for account, job and vendor

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html [hunk 14]
```diff
@@ -170,5 +171,58 @@ <h5 *ngIf="!isCloneSettingView" class="!tw-text-text-muted tw-text-14-px !tw-fon
...
+            </div>
+        </div>
+        <div>
+            <div *ngFor="let defaultField of defaultImportFields">
+                <div *ngIf="form.controls[defaultField.formController].value && defaultField.import_code">
+                    <app-configuration-select-field
+                    [appName]="appName"
+                    [form]="form"
+                    [options]="importCodeSelectorOptions"
+                    [placeholder]="'Select the how to import'"
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html:200]: placeholder seems incorrect - if we select 1 option by default then need not send placeholder

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html [hunk 15]
```diff
@@ -170,5 +171,58 @@ <h5 *ngIf="!isCloneSettingView" class="!tw-text-text-muted tw-text-14-px !tw-fon
...
+                    [label]="'How to Import the '+ helper.sentenseCaseConversion(defaultField.destination_field) +' from ' + appName + '?'"
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html:201]: make it small case, and all sentence case

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html [hunk 16]
```diff
@@ -170,5 +171,58 @@ <h5 *ngIf="!isCloneSettingView" class="!tw-text-text-muted tw-text-14-px !tw-fon
...
+                    [subLabel]="'You have the option to import either just the names of your '+ helper.sentenseCaseConversion(defaultField.destination_field) +' or both the names and codes. Choose the option that best suits how your employees identify the values.'"
+                    [isMultiLineOption]="true"
+                    [isDisabled]="form.controls[defaultField.formController].value && !isOnboarding"
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html:204]: even if they have saved the setting earlier, it should be disabled

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html [hunk 17]
```diff
@@ -170,5 +171,58 @@ <h5 *ngIf="!isCloneSettingView" class="!tw-text-text-muted tw-text-14-px !tw-fon
...
+            </div>
+            <div formArrayName="expenseFields">
+                <div *ngFor="let expenseField of expenseFieldsGetter.controls; let i = index" [formGroupName]="i">
+                    <div *ngIf="expenseField.value.import_to_fyle">
+                        <app-configuration-select-field
+                        [appName]="appName"
+                        [form]="getFormGroup(expenseField)"
+                        [options]="importCodeSelectorOptions"
+                        [placeholder]="'Select the how to import'"
+                        [label]="'How to Import the '+ helper.sentenseCaseConversion(expenseField.value.destination_field) +' from ' + appName + '?'"
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html:217]: same as above

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.scss [hunk 18]
```diff
@@ -133,4 +133,8 @@
 .container .checkmark:after {
   content: url('./../../../../../assets/icons/co/grv-checkmark-small.svg') !important;
   @apply tw-left-[3px] tw-top-0 tw-w-6-px tw-h-10-px tw-border-t-0-px tw-border-r-0-px tw-border-b-0-px tw-border-l-0-px tw-border-white;
+}
+
+:host ::ng-deep app-configuration-select-field > div > div > div > div{
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.scss:138]: bad practice, apply in html

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts [hunk 19]
```diff
@@ -123,8 +149,6 @@ export class ConfigurationImportFieldComponent implements OnInit {
 
       // Get the 'import_to_fyle' control at the specified index and disable it
       (this.form.get('expenseFields') as FormArray).at(index)?.get('import_to_fyle')?.disable();
-    } else {
-      (this.form.get('expenseFields') as FormArray).at(index)?.get('import_to_fyle')?.setValue(true);
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts:127]: why are we removing this? we should enable toggle by default when dropdown value is changed

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts [hunk 20]
```diff
@@ -167,6 +191,11 @@ export class ConfigurationImportFieldComponent implements OnInit {
   }
 
   onShowWarningForDependentFields(event: any, formGroup: AbstractControl): void {
+    if (event.checked && this.appName === AppName.SAGE300) {
+      this.importCodeEnabled.push(true);
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts:195]: why is this an array?

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts [hunk 21]
```diff
@@ -167,6 +191,11 @@ export class ConfigurationImportFieldComponent implements OnInit {
...
+    } else if (!event.checked && this.appName === AppName.SAGE300) {
+      this.importCodeEnabled.pop();
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts:197]: toggle is present for dependent fields too, why are we pushing or popping for those changes too?

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts [hunk 22]
```diff
@@ -183,6 +212,16 @@ export class ConfigurationImportFieldComponent implements OnInit {
...
+  ngOnInit(): void {
+    this.isOnboarding = this.workspace.getOnboardingState() === 'IMPORT_SETTINGS';
+    Object.keys(this.form.controls).forEach(key => {
+      if (key !== 'expenseFields' && key !== 'dependentFieldImportToggle') {
+        this.form.get(key)?.value === true ? this.importCodeEnabled.push(true) : '';
+      }
+    });
+    Object.keys(this.expenseFieldsGetter.controls).forEach(key => {
+      this.expenseFieldsGetter.controls[(key as unknown as number)].get('import_code')?.value === true ? this.importCodeEnabled.push(true) : '';
+    });
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts:224]: what are we trying to do here

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html [hunk 23]
```diff
@@ -76,7 +78,7 @@ <h5 *ngIf="formControllerName === 'cccExportType' && brandingConfig.brandId ===
                         <span *ngIf="brandingConfig.brandId !== 'co'"> how it looks on {{appName}}</span>
                         <span *ngIf="brandingConfig.brandId === 'co'"> in {{appName}}</span>
                     </div>
-                    <div *ngIf="!showExportPreview && isDisabled" class="sub-text tw-text-pretty">
+                    <div *ngIf="!showExportPreview && isDisabled && !formControllerName.includes('Code') && formControllerName !== 'import_code'" class="sub-text tw-text-pretty">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:81]: let's have proper input for this and not add hacks

File: src/app/core/models/common/import-settings.model.ts (hunk 3 above)
Comment (DhaaraniCIT) [reply src/app/core/models/common/import-settings.model.ts:89]: to add `import_code` field in fieldDate

File: src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts (hunk 5 above)
Comment (DhaaraniCIT) [reply src/app/integrations/sage300/sage300-shared/sage300-import-settings/sage300-import-settings.component.ts:70]: this is the default fields ka structure for category and vendor
Screenshot 2024-08-02 at 11 09 35 AM

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts (hunk 20 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts:195]: this array is to have a check, weather to show the import code field section or not, when the array has at least one entry it will show the section else it won't

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts (hunk 21 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts:197]: this pushing and poping to have tract weather we have a requirement to show the import code field or not

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts (hunk 22 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts:224]: this is for showing the import code section after one save or in post onboarding and after the import settings save in onboarding

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html (hunk 14 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html:200]: no default values for import code and also add dynamic validation, separate PR

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts [hunk 24]
```diff
@@ -95,6 +95,10 @@ export class ConfigurationImportFieldComponent implements OnInit {
     return this.form.get('expenseFields') as FormArray;
   }
 
+  disabledToolTipText(expenseField: { value: { source_field: any; }; }): string {
+    return !expenseField.value.source_field ? this.helper.sentenseCaseConversion('Fill in the Fyle Field toggle will be enabled') : '';
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.ts:99]: pls get content from Kamalini -  doesn't sound right

File: src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html [hunk 25]
```diff
@@ -194,32 +194,34 @@ <h4 class="tw-text-16-px tw-text-normal-text-color tw-mb-8-px">
...
-                        [label]="'How to Import the '+ helper.sentenseCaseConversion(expenseField.value.destination_field) +' from ' + appName + '?'"
-                        [subLabel]="'You have the option to import either just the names of your '+ helper.sentenseCaseConversion(expenseField.value.destination_field) +' or both the names and codes. Choose the option that best suits how your employees identify the values.'"
-                        [isMultiLineOption]="true"
-                        [isDisabled]="expenseField.value.import_to_fyle && !isOnboarding"
-                        [formControllerName]="'import_code'"
-                        ></app-configuration-select-field>
+                            [appName]="appName"
+                            [isDisableTextRequired]="false"
+                            [form]="getFormGroup(expenseField)"
+                            [options]="importCodeSelectorOptions"
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-import-field/configuration-import-field.component.html:217]: need to be updated

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html [hunk 26]
```diff
@@ -78,7 +78,7 @@ <h5 *ngIf="formControllerName === 'cccExportType' && brandingConfig.brandId ===
                         <span *ngIf="brandingConfig.brandId !== 'co'"> how it looks on {{appName}}</span>
                         <span *ngIf="brandingConfig.brandId === 'co'"> in {{appName}}</span>
                     </div>
-                    <div *ngIf="!showExportPreview && isDisabled && !formControllerName.includes('Code') && formControllerName !== 'import_code'" class="sub-text tw-text-pretty">
+                    <div *ngIf="!showExportPreview && isDisabled && isDisableTextRequired" class="sub-text tw-text-pretty">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:81]: I don't find any file where we're sending isDisableTextRequired as true, we might end up not displaying this anywhere - can we find out where is this used earlier

Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:81]: by default, it will be true, if we don't want we can send false

File: src/app/core/models/common/import-settings.model.ts [hunk 27]
```diff
@@ -52,11 +54,19 @@ export class ImportSettingsModel {
...
       is_custom: new FormControl(data.is_custom || false),
-      source_placeholder: new FormControl(data.source_placeholder || null)
+      source_placeholder: new FormControl(data.source_placeholder || null),
+      import_code: new FormControl(data.import_code || false)
     });
   }
 
-  static constructFormArray(importSettingsMappingSettings: ImportSettingMappingRow[] | [], accountingAppFields: IntegrationField[], isDestinationFixedImport: boolean = true): FormGroup[] {
+  static getImportCodeField(importCodeFields: string[], destinationField: string): boolean {
+    if (importCodeFields?.length) {
```
Comment (ashwin1111) [inline src/app/core/models/common/import-settings.model.ts:63]: if not required

**Review by ashwin1111 — APPROVED** [review]
